tests/integration_tests/security/test_vulnerabilities.py

```diff
@@ -11,55 +11,140 @@
 import pytest
 import requests
 
-from framework import utils
-from framework.ab_test import git_ab_test_command_if_pr, set_did_not_grow_comparator
+from framework.ab_test import (
+    git_ab_test_command_if_pr,
+    git_ab_test_ssh_command,
+    git_ab_test_ssh_command_if_pr,
+    is_pr,
+    set_did_not_grow_comparator,
+)
 from framework.properties import global_props
 from framework.utils import CommandReturn
 from framework.utils_cpu_templates import nonci_on_arm
 
 CHECKER_URL = "https://meltdown.ovh"
 CHECKER_FILENAME = "spectre-meltdown-checker.sh"
+REMOTE_CHECKER_PATH = f"/tmp/{CHECKER_FILENAME}"
+REMOTE_CHECKER_COMMAND = (
+    f"sh {REMOTE_CHECKER_PATH} --explain --no-intel-db --batch json"
+)
+
+VULN_DIR = "/sys/devices/system/cpu/vulnerabilities"
 
+# This lint doesnt work well with fixtures
+# pylint:disable=redefined-outer-name
 
-@pytest.fixture(name="microvm")
-def microvm_fxt(uvm_plain):
-    """Microvm fixture"""
-    uvm_plain.spawn()
-    uvm_plain.basic_config(
-        vcpu_count=2,
-        mem_size_mib=256,
+
+def configure_microvm(
+    factory,
+    kernel,
+    rootfs,
+    *,
+    firecracker=None,
+    jailer=None,
+    cpu_template=None,
+    custom_cpu_template=None,
+):
+    """Build a microvm for vulnerability tests"""
+    assert not (cpu_template and custom_cpu_template)
+    # Either both or neither are specified
+    assert firecracker and jailer or not firecracker and not jailer
+
+    if firecracker:
+        microvm = factory.build(
+            kernel, rootfs, fc_binary_path=firecracker, jailer_binary_path=jailer
+        )
+    else:
+        microvm = factory.build(kernel, rootfs)
+
+    microvm.spawn()
+    microvm.basic_config(vcpu_count=2, mem_size_mib=256, cpu_template=cpu_template)
+    if custom_cpu_template:
+        microvm.api.cpu_config.put(**custom_cpu_template["template"])
+    microvm.cpu_template = cpu_template or custom_cpu_template
+    microvm.add_net_iface()
+    microvm.start()
+    return microvm
+
+
+@pytest.fixture
+def build_microvm(
+    microvm_factory,
+    guest_kernel_linux_5_10,
+    rootfs_ubuntu_22,
+):
+    """Fixture returning a factory function for a normal microvm"""
+    return lambda firecracker=None, jailer=None: configure_microvm(
+        microvm_factory,
+        guest_kernel_linux_5_10,
+        rootfs_ubuntu_22,
+        firecracker=firecracker,
+        jailer=jailer,
     )
```

Comment:
The pattern here in the `test: ab-ify test_vulnerabilities.py` commit I find quite difficult to follow. We may be able to mimic this functionality with a builder pattern that could be a simpler.

Reply:
Tbh, I'm not the biggest fan of putting a builder on top of the `MicroVMFactory` thing. It doesn't change much, but it does make the entire thing stateful and give us a huge `.build()` method that will essentially combine the code that's currently split across `with_restore`, `with_checker` and `configure_microvm`

Reply:
Okay.